### PR TITLE
adding openai/triton

### PR DESCRIPTION
**Pull request by erip**
cc @conda-forge/help-python-c

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by erip** [comment]
Note: triton relies on static libraries -- will need to figure out how best to handle that in the condification process

**Comment by h-vetinari** [comment]
This is gonna have to wait for https://github.com/conda-forge/staged-recipes/pull/17303, IMO (which I started to get to triton...)

**Comment by erip** [comment]
@h-vetinari yes, this will definitely need to wait for some of the other deps to land. I thought I'd get some of the grunt work out of the way. Shall I add you as a maintainer here, too?

**Comment by h-vetinari** [comment]
> @h-vetinari yes, this will definitely need to wait for some of the other deps to land. I thought I'd get some of the grunt work out of the way. Shall I add you as a maintainer here, too?

I think most things are there, even https://github.com/conda-forge/dlfcn-win32-feedstock. The more involved thing will be to patch out the dependence on static libraries (as well as a lot of the hyper-specific LLVM-version stuff).

You could add me as a collaborator to your fork of staged-recipes, then we could work on this PR together. 🙃

**Comment by h-vetinari** [comment]
Finally https://github.com/conda-forge/staged-recipes/pull/17303 is merged 🥳 

Will be coming back to this once the builds are up and I find some time.

**Comment by h-vetinari** [comment]
@conda-forge/staged-recipes 
I think this is ready for a first look**. It's compiling against the static LLVM libs, but those are [packaged](https://github.com/regro/libcfgraph/blob/master/artifacts/llvmdev/conda-forge/linux-64/llvmdev-11.1.0-hf817b99_3.json#L1790-L1940) in `llvmdev`, and shouldn't make it into the final package.

I'm not sure yet if we should package CPU (i.e. non-CUDA) builds. Upstream [says](https://github.com/openai/triton/tree/v1.1.2#compatibility) they're still under development (though conceivably, the failing test due to not finding `nvidia-smi` could be skipped.

With some more intrusive patching, we could also package the C++ libs+headers separately into a `libtriton`, not sure if there are consumers for this.

** It might still need some light touch ups, but I'm leaving for the weekend, and the builds pass far enough to clear the import and start running the test suite.

**Comment by h-vetinari** [comment]
@conda-forge/cudatoolkit 

This is running into a missing `nvidia-smi` binary. Should this be available somewhere (searching conda-forge only brings up the gputils-[feedstock](https://github.com/conda-forge/gputil-feedstock), but that's just a python wrapper...)? I tested cudatoolkit as well, but no cigar.

If it's not possible to run this in our CI, I'll have to patch the upstream test suite it seems.

**Comment by h-vetinari** [comment]
@conda-forge/staged-recipes 
This one should be ready now (I cannot remove the "WIP" tag because this PR was opened by @erip, who I'm collaborating with).

Unfortunately we cannot run the upstream test suite as it absolutely requires a physical GPU to run. I've tried patching this out, but to no avail, and upstream [confirmed](https://github.com/openai/triton/issues/466#issuecomment-1058367541) that CPU-only isn't possible yet _at all_.

The cutlass bindings have also been removed as they are broken and effectively [considered](https://github.com/openai/triton/issues/466#issuecomment-1057654719) obsolete by upstream.

CC @ptillet; you'd be very welcome to participate in maintaining this recipe if you want. 🙃
As a sidenote, I'm pretty sure that the `cuda.h` you're vendoring is not license-conforming. I've worked around this here.

PS. The failure on CUDA 11.1 was a flake due to `conda.CondaMultiError: [Errno 28] No space left on device`

**Comment by h-vetinari** [comment]
@erip
Can you please remove the WIP tag?

**Comment by h-vetinari** [comment]
@conda-forge/staged-recipes 
Could someone PTAL? 🙃

**Comment by ocefpaf** [comment]
Do we have a cuda team? I don't feel comfortable merging this b/c I don't really understand cuda. Looks fine to me, so ping me again if an expert doesn't show up in the next few days and we'll see what we can do to get this in.

**Comment by jaimergp** [comment]
> Do we have a cuda team? 

I guess the closer we have is nvcc / cudatoolkit teams.

> This is running into a missing nvidia-smi binary

@h-vetinari, I think this is only distributed with the drivers :/

**Comment by jaimergp** [comment]
I'd also take a look at this [recent recipe](https://github.com/conda-forge/nnpops-feedstock/blob/main/recipe/meta.yaml) for the Pytorch stuff. @hmaarrfk might have some more input too!

**Comment by h-vetinari** [comment]
Thanks @jaimergp!

I've been working on a couple pytorch-related feedstocks, so I know the drill for that by now. ;-)
This recipe does not have pytorch as a host-dep (and only builds for CUDA), so we can avoid a lot of the otherwise necessary ceremony.

**Comment by h-vetinari** [comment]
> > Do we have a cuda team?
> 
> I guess the closest we have is nvcc / cudatoolkit teams.

@conda-forge/cudatoolkit, could someone help out here with review here, please? 🙃 

This is required to build xformers and has been sitting in the queue for a while.

**Comment by leofang** [comment]
I can take a look tonight!

File: recipes/triton/meta.yaml
Comment (leofang) [inline recipes/triton/meta.yaml:37]: Couldn't we just list `pytorch-gpu` here?

File: recipes/triton/meta.yaml
Comment (leofang) [inline recipes/triton/meta.yaml:20]: I don't understand. Why do we have a cbc yaml file including OS X targets, while this recipe requires CUDA? macOS has no CUDA support these days.

**Review by leofang — COMMENTED** [review]
Mostly LGTM. Left a few questions. 

Another question is none of the CUDA libraries is linked by Triton, so we see this warning:
```
WARNING (triton): dso library package conda-forge::cudatoolkit-11.0.3-h15472ef_10 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
```
@h-vetinari, do you know is this because CUDA binaries are dlopen'd at runtime by Triton?

File: recipes/triton/meta.yaml
Comment (h-vetinari) [reply recipes/triton/meta.yaml:20]: The CBC was an artefact from trying to get the CPU builds to work until I found out that upstream does not support that at all at the moment - will remove.

File: recipes/triton/meta.yaml
Comment (h-vetinari) [reply recipes/triton/meta.yaml:37]: Because that package is not used anymore by the `pytorch` channel, and it was only a compatibility wrapper on the side of conda-forge as well. CC @hmaarrfk

**Comment by h-vetinari** [comment]
> Mostly LGTM. Left a few questions.

Thanks for taking a look! :)

> Another question is none of the CUDA libraries is linked by Triton, so we see this warning:
> 
> ```
> WARNING (triton): dso library package conda-forge::cudatoolkit-11.0.3-h15472ef_10 in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
> ```
> 
> @h-vetinari, do you know is this because CUDA binaries are dlopen'd at runtime by Triton?

It might be because we're not fully testing this in CI, because the package is so dependent on a physical GPU being present. I doubt that it's completely unused, but admit to not knowing this definitely. Maybe @ptillet can comment?

**Comment by ptillet** [comment]
First of all, thanks guys for all the work :)  @h-vetinari  Yep, CUDA binaries (specifically libcuda.so) is dlopen'ed at runtime. Triton also calls `ptxas` to compile GPU code. As long as the user has the proprietary GPU driver on their machine (libcuda.so) and `ptxas` in their path, Triton should work as expected.

File: recipes/triton/meta.yaml
Comment (leofang) [reply recipes/triton/meta.yaml:37]: hmmm I thought conda-forge aims to be self-contained so I don't understand why we care about other channels, though I can see this increases compatibility. Can someone from @conda-forge/staged-recipes confirm?

Comment (hmaarrfk) [reply recipes/triton/meta.yaml:37]: We do aim to be self contained. But the the truth is we do benefit from following upstream where we can.

Leaving this syntax would benefit users building the windows versions for themselves

**Comment by ptillet** [comment]
@leofang  Triton 1.x uses the JIT included in `libcuda.so` when `ptxas` is not present in the system, so you should be good. Note, however, that this behavior is however currently deprecated in `master` because it can lead to deceptively low performance for people who have a recent CUDA on top of an old-ish driver -- the CUDA compilers have been improving a whole lot since version 11.3. This is easy to reinstantiate previous behavior, though.

Is `nvcc` really not in `cudatoolkit-dev` ? I always thought it was :sweat_smile.

File: recipes/triton/meta.yaml
Comment (leofang) [reply recipes/triton/meta.yaml:37]: Sounds good, thanks Mark. Let's leave it as is.

**Review by leofang — APPROVED** [review]
LGTM!

File: recipes/triton/meta.yaml
Comment (carterbox) [inline recipes/triton/meta.yaml:29]:
```suggestion
  host:
    - cudatoolkit {{ cuda_compiler_version }}
```
Because the `nvcc=11.2` compiler has a run_export for `cudatoolkit>=11.2,<12` instead of `cudatoolkit=11.2` your [11.2 build has a runtime dependency on `cudatoolkit=11.6`](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=485510&view=logs&j=eb6ecf2b-0363-5c1a-baa7-2d771889d197&t=d8576026-8998-52ff-d027-1d631b0fb837&l=4296)! I think we need to specify the toolkit version so that the solver uses the correct version in the build environment and then users get the correct toolkit.

@leofang, is this the correct way to deal with this? Or should we be setting a toolkit requirement that matches the nvcc requirement? i.e. `cudatoolkit>=11.2,<12` instead of `cudatoolkit=11.2`

More reading: https://github.com/conda-forge/nvcc-feedstock/pull/71

**Review by carterbox — CHANGES_REQUESTED** [review]

File: recipes/triton/meta.yaml
Comment (h-vetinari) [reply recipes/triton/meta.yaml:29]: That's intentional, and the PR you linked is the background.

> your 11.2 build has a runtime dependency on `cudatoolkit=11.6`

No, it's not a runtime dependence on 11.6, it's a run-constraint on >=11.2,<12 (which means that the appropriate cudatoolkit can be chosen based on the installation in the users environment, c.f. the `__cuda` virtual package).

Comment (carterbox) [reply recipes/triton/meta.yaml:29]: Nevermind, the run export is also `cudatoolkit>=11.2,12`, so users won't have a problem, but it's weird that out build environment is choosing the `cudatoolkit=11.6` package.

Comment (leofang) [reply recipes/triton/meta.yaml:29]: I think it's normal that we build with cuda 11.2 and run with 11.6.

**Comment by h-vetinari** [comment]
Thanks. Considering the work that went into this, it would have been nice to keep some `Coauthored-By` tags (or indeed, set the author correctly and stay committer only).

**Comment by leofang** [comment]
Ah, yes, no squash & merge.

**Comment by h-vetinari** [comment]
> Ah, yes, no squash & merge.

Not sure what the policy is on staged-recipes; I'm not against squashes (and it's not like all 51 commits here needed to be kept), but would then at least maintain the tags (that GH suggests AFAIK) for proper attribution.

**Comment by leofang** [comment]
Yup though it's harder to retain the attributes when squashing. Some manual work is required, and it's easier to not squash.

**Comment by jakirkham** [comment]
However it was not squash merged

**Comment by h-vetinari** [comment]
Not trying to make a big deal out of this BTW, just thought I'd mention it as something to improve next time.

**Comment by leofang** [comment]
John is right, sorry I overlooked. The commit history of this repo is long so we need to scroll to the next page (as of now).

**Comment by leofang** [comment]
btw the feedstock is still empty? https://github.com/conda-forge/triton-feedstock

**Comment by jakirkham** [comment]
It takes time for things to convert. Sometimes we get rate limited or a CI's API is flaky or some other issue. We autoretry periodically. Would just give it time 🙂

**Comment by h-vetinari** [comment]
> However it was not squash merged

Hm, that's on me then, apologies @carterbox. I had (mis-)remembered that merge commits don't show the diff (fully, correctly or sometimes at all; not sure if GH's UI improved on that recently).

**Comment by carterbox** [comment]
@h-vetinari, squash-merge is not enabled in this repo (staged-recipes). Also, anecdotally, the last time I tried to squash/rebase someone else's PR to clean up the history, GitHub closed the PR. 😅

I agree that it's a bit of a shame that author info is lost, but it's all lost when the recipe is copied into it's own feedstock.

**Pull request merged**
